File: app/lib/environment.rb
# frozen_string_literal: true

module Environment
  TRUTHY_VALUES = %w[1 true TRUE yes YES].freeze
  FALSY_VALUES = %w[0 false FALSE no NO].freeze

  # Errors
  Error = Class.new(StandardError)
  KeyNotFoundError = Class.new(Error)

  class << self
    def self_hosted?(env: ENV)
      truthy?(env['TUIST_CLOUD_SELF_HOSTED'])
    end

    def tuist_hosted?(env: ENV)
      !self_hosted?(env: env)
    end

    def production_like_env?(env: ENV)
      (Rails.env.production? || Rails.env.staging? || Rails.env.canary?) && !precompiling_assets?(env: env)
    end

    def truthy?(value)
      return false if value.blank?

      TRUTHY_VALUES.any? { |v| v == value.to_s }
    end

    def falsy?(value)
      return true if value.blank?

      FALSY_VALUES.any? { |v| v == value.to_s }
    end

    def fetch(*args, env: ENV, credentials: Rails.application.credentials, defaults: Rails.application.config.defaults,
      with_prefix: true)
      env_variable_key = args.join('_').upcase.to_s
      env_variable_key = "TUIST_#{env_variable_key}" if with_prefix
      env_variable_value = env.to_h.fetch(env_variable_key, nil)
      credentials_value = credentials.dig(*args)
      defaults_value = defaults.dig(*args)
      env_variable_value || credentials_value || defaults_value
    end

    def fetch!(*args, env: ENV, credentials: Rails.application.credentials, defaults: Rails.application.config.defaults)
      value = fetch(*args, env:, credentials:, defaults:)
      if value.blank?
        raise KeyNotFoundError, "The key #{args.map(&:to_s).join('.')} was not found in the app credentials"
      end

      value
    end

    # Getters

    def database_url
      fetch(:database_url, with_prefix: false)
    end

    def aws_access_key_id
      fetch(:aws, :access_key_id)
    end

    def aws_secret_access_key
      fetch(:aws, :secret_access_key)
    end

    def aws_region
      fetch(:aws, :region)
    end

    def aws_endpoint
      fetch(:aws, :endpoint)
    end

    def aws_bucket_name
      fetch(:aws, :bucket_name)
    end

    def storage_configured?
      aws_configured?
    end

    def blocklisted_slug_keywords
      fetch(:blocklisted_slug_keywords)
    end

    def secret_key_tokens
      fetch(:secret_key, :tokens) || secret_key_base
    end

    def secret_key_password
      fetch(:secret_key, :password) || secret_key_base
    end

    def secret_key_base(env: ENV, credentials: Rails.application.credentials,
      defaults: Rails.application.config.defaults, with_prefix: true)
      fetch(:secret_key, :base, env: env, credentials: credentials, defaults: defaults, with_prefix: with_prefix)
    end

    def app_url
      fetch(:app_url)
    end

    def github_oauth_id
      fetch(:github, :oauth_id)
    end

    def github_oauth_secret
      fetch(:github, :oauth_secret)
    end

    def okta_site
      fetch(:okta, :site)
    end

    def okta_client_id
      fetch(:okta, :client_id)
    end

    def okta_client_secret
      fetch(:okta, :client_secret)
    end

    def okta_authorize_url
      fetch(:okta, :authorize_url)
    end

    def okta_token_url
      fetch(:okta, :token_url)
    end

    def okta_user_info_url
      fetch(:okta, :user_info_url)
    end

    def attio_api_key
      fetch(:attio, :api_key)
    end

    def stripe_api_key
      fetch(:stripe, :secret_key)
    end

    def stripe_publishable_key
      fetch(:stripe, :publishable_key)
    end

    def stripe_endpoint_secret
      fetch(:stripe, :endpoint_secret)
    end

    def precompiling_assets?(env: ENV)
      truthy?(env['SECRET_KEY_BASE_DUMMY'])
    end

    def smpt_domain
      fetch(:smpt_settings, :domain)
    end

    def smpt_user_name
      fetch(:smpt_settings, :user_name)
    end

    def smpt_password
      fetch(:smpt_settings, :password)
    end

    # Configuration checkers

    def attio_configured?
      attio_api_key.present?
    end

    def stripe_configured?
      stripe_api_key.present? &&
        stripe_publishable_key.present? &&
        stripe_endpoint_secret.present?
    end

    def smpt_configured?
      smpt_domain.present? && smpt_user_name.present? && smpt_password.present?
    end

    def okta_configured?
      okta_site.present? && okta_client_id.present? && okta_client_secret.present?
    end

    def github_configured?
      github_oauth_id.present? && github_oauth_secret.present?
    end

    def aws_configured?
      aws_access_key_id.present? && aws_secret_access_key.present? && aws_region.present? &&
        (tuist_hosted? || aws_bucket_name.present?)
    end

    def app_url_configured?
      app_url.present?
    end

    def database_configured?
      database_url.present?
    end

    def secret_key_base_configured?
      secret_key_base.present?
    end

    def ensure_configured!
      return if Rails.env.test? || Rails.env.development? || precompiling_assets?

      errors = []
      errors << "Storage is not configured" unless storage_configured?
      errors << "Application URL is not configured" unless app_url_configured?
      errors << "Database is not configured" unless database_configured?
      errors << "Secret key base is not configured" unless secret_key_base_configured?
      unless Environment.self_hosted?
        errors << "Stripe is not configured" unless stripe_configured?
      end

      if errors.any?
        raise Error, <<~ERROR
          Can't start Tuist Cloud due to the following errors:
          #{errors.map { |error| " - #{error}" }.join("\n")}

          Please, check our documentation to learn how to configure Tuist Cloud: https://github.com/tuist/cloud-enterprise
        ERROR
      end
    end
  end
end

File: lib/tuist/environment.ex
defmodule Tuist.Environment do
  @moduledoc false
  @env Mix.env()

  defmodule Version do
    @moduledoc ~S"""
    A module that represents a Tuist version.
    Tuist versions follow the convention MAJOR.YY.MM.DD.
    """
    @type t :: %{
            major: integer(),
            date: Date.t()
          }
    @enforce_keys [:major, :date]

    defstruct [:major, :date]
  end

  def env, do: @env

  def test?() do
    env() == :test
  end

  defmacro run_if_error_tracking_enabled(block) do
    enabled = error_tracking_enabled?()

    if enabled do
      block
    else
      quote do
        :ok
      end
    end
  end

  def truthy?(value) do
    ["1", "true", "TRUE", "yes", "YES"] |> Enum.member?(value)
  end

  def on_premise?() do
    not truthy?(System.get_env("TUIST_CLOUD_HOSTED", "0")) and
      not truthy?(System.get_env("TUIST_HOSTED", "0"))
  end

  def log_level() do
    System.get_env("TUIST_LOG_LEVEL", "info") |> String.to_atom()
  end

  def use_ssl_for_database?() do
    truthy?(System.get_env("TUIST_USE_SSL_FOR_DATABASE", "1"))
  end

  def use_ipv6?(secrets \\ secrets()) do
    get([:use_ipv6], secrets)
  end

  def database_pool_size(secrets \\ secrets()) do
    case get([:database, :pool_size], secrets) do
      pool_size when is_binary(pool_size) -> String.to_integer(pool_size)
      _ -> 10
    end
  end

  def license_features() do
    Application.get_env(:tuist, :license).features
  end

  def analytics_enabled?() do
    not on_premise?() and env() == :prod
  end

  def error_tracking_enabled?() do
    not on_premise?() and Enum.member?([:prod, :stag, :can], env())
  end

  def version() do
    version = get([:version])

    case version do
      nil ->
        %Version{major: "1", date: Date.utc_today()}

      "" ->
        %Version{major: "1", date: Date.utc_today()}

      version ->
        [major, yy, mm, dd] = String.split(version, ".")

        date =
          Date.from_iso8601("#{yy}-#{mm}-#{dd}")
          |> case do
            {:ok, date} -> date
            # Fallback in case of error
            {:error, _} -> Date.utc_today()
          end

        %Version{major: major, date: date}
    end
  end

  def attio_api_key(secrets \\ secrets()) do
    get([:attio, :api_key], secrets)
  end

  def posthog_api_key(secrets \\ secrets()) do
    get([:posthog, :api_key], secrets)
  end

  def posthog_url(secrets \\ secrets()) do
    get([:posthog, :url], secrets)
  end

  def s3_access_key_id(secrets \\ secrets()) do
    System.get_env("AWS_ACCESS_KEY_ID") || get([:aws, :access_key_id], secrets) ||
      get([:s3, :access_key_id], secrets)
  end

  def s3_secret_access_key(secrets \\ secrets()) do
    System.get_env("AWS_SECRET_ACCESS_KEY") || get([:aws, :secret_access_key], secrets) ||
      get([:s3, :secret_access_key], secrets)
  end

  def aws_region(secrets \\ secrets()) do
    System.get_env("AWS_REGION") || get([:aws, :region], secrets) || "auto"
  end

  def aws_session_token(secrets \\ secrets()) do
    System.get_env("AWS_SESSION_TOKEN") || get([:aws, :session_token], secrets)
  end

  def aws_profile(secrets \\ secrets()) do
    System.get_env("AWS_PROFILE") || get([:aws, :profile], secrets)
  end

  def aws_use_session_token?(secrets \\ secrets()) do
    get([:aws, :use_session_token], secrets)
  end

  def s3_region(secrets \\ secrets()) do
    get([:aws, :region], secrets) || get([:s3, :region], secrets)
  end

  def s3_bucket_name(secrets \\ secrets()) do
    get([:aws, :bucket_name], secrets) || get([:s3, :bucket_name], secrets)
  end

  def s3_endpoint(secrets \\ secrets()) do
    get([:aws, :endpoint], secrets) || get([:s3, :endpoint], secrets)
  end

  def s3_pool_size(secrets \\ secrets()) do
    case get([:s3, :pool_size], secrets) do
      pool_size when is_binary(pool_size) -> String.to_integer(pool_size)
      # Since we use http2, which allows multi-plexing, the size of the pool can be smaller
      _ -> 500
    end
  end

  def s3_pool_count(secrets \\ secrets()) do
    case get([:s3, :pool_count], secrets) do
      pool_count when is_binary(pool_count) -> String.to_integer(pool_count)
      _ -> 1
    end
  end

  def s3_protocol(secrets \\ secrets()) do
    case get([:s3, :protocol], secrets) do
      protocol when is_binary(protocol) -> protocol |> String.to_atom()
      _ -> :http1
    end
  end

  def s3_configured?(secrets \\ secrets()) do
    s3_access_key_id(secrets) != nil and s3_secret_access_key(secrets) != nil and
      s3_region(secrets) != nil and
      (!on_premise?() or s3_bucket_name(secrets) != nil)
  end

  def slack_tuist_token(secrets \\ secrets()) do
    get([:slack, :tuist, :token], secrets)
  end

  def stripe_api_key(secrets \\ secrets()) do
    get([:stripe, :secret_key], secrets)
  end

  def stripe_publishable_key(secrets \\ secrets()) do
    get([:stripe, :publishable_key], secrets)
  end

  def stripe_endpoint_secret(secrets \\ secrets()) do
    get([:stripe, :endpoint_secret], secrets)
  end

  def stripe_configured?(secrets \\ secrets()) do
    stripe_api_key(secrets) != nil and stripe_publishable_key(secrets) != nil and
      stripe_endpoint_secret(secrets) != nil
  end

  def stripe_prices(secrets \\ secrets()) do
    get_in(secrets, [:stripe, :prices])
  end

  def github_app_client_id(secrets \\ secrets()) do
    get([:github, :app_client_id], secrets) || get([:github, :oauth_id], secrets)
  end

  def github_app_client_secret(secrets \\ secrets()) do
    get([:github, :app_client_secret], secrets) || get([:github, :oauth_secret], secrets)
  end

  def github_app_private_key(secrets \\ secrets()) do
    get([:github, :app_private_key], secrets)
  end

  def github_auth_configured?(secrets \\ secrets()) do
    github_app_client_id(secrets) != nil and github_app_client_secret(secrets) != nil
  end

  def github_app_configured?(secrets \\ secrets()) do
    github_auth_configured?(secrets) and github_app_private_key(secrets) != nil
  end

  def google_oauth_client_id(secrets \\ secrets()) do
    get([:google, :oauth_client_id], secrets)
  end

  def google_oauth_client_secret(secrets \\ secrets()) do
    get([:google, :oauth_client_secret], secrets)
  end

  def google_oauth_configured?(secrets \\ secrets()) do
    google_oauth_client_id(secrets) != nil and google_oauth_client_secret(secrets) != nil
  end

  def okta_site(secrets \\ secrets()) do
    get([:okta, :site], secrets)
  end

  def okta_client_id(secrets \\ secrets()) do
    get([:okta, :client_id], secrets)
  end

  def okta_client_secret(secrets \\ secrets()) do
    get([:okta, :client_secret], secrets)
  end

  def okta_authorize_url(secrets \\ secrets()) do
    get([:okta, :authorize_url], secrets)
  end

  def okta_token_url(secrets \\ secrets()) do
    get([:okta, :token_url], secrets)
  end

  def okta_user_info_url(secrets \\ secrets()) do
    get([:okta, :user_info_url], secrets)
  end

  def okta_configured?(secrets \\ secrets()) do
    okta_site(secrets) != nil and okta_client_id(secrets) != nil and
      okta_client_secret(secrets) != nil
  end

  def mailgun_api_key(secrets \\ secrets()) do
    get([:mailgun, :api_key], secrets)
  end

  def smtp_domain(secrets \\ secrets()) do
    get([:smtp_settings, :domain], secrets)
  end

  def smtp_user_name(secrets \\ secrets()) do
    get([:smtp_settings, :user_name], secrets)
  end

  def mail_configured?(secrets \\ secrets()) do
    mailgun_api_key(secrets) != nil and smtp_domain(secrets) != nil and
      smtp_user_name(secrets) != nil
  end

  def app_url(secrets \\ secrets(), opts \\ []) do
    path = Keyword.get(opts, :path, "/")
    url = get([:app, :url], secrets) || "http://localhost:8080"
    %{URI.parse(url) | path: path} |> URI.to_string()
  end

  def app_signal_push_api_key(secrets \\ secrets()) do
    get([:app_signal, :push_api_key], secrets)
  end

  def secret_key_base(secrets \\ secrets()) do
    get([:secret_key, :base], secrets)
  end

  def secret_key_password(secrets \\ secrets()) do
    get([:secret_key, :password], secrets, default_value: secret_key_base(secrets))
  end

  def secret_key_tokens(secrets \\ secrets()) do
    get([:secret_key, :tokens], secrets, default_value: secret_key_base(secrets))
  end

  def get(keys, secrets \\ secrets(), opts \\ []) do
    env_variable =
      "TUIST_#{keys |> Enum.map(&Atom.to_string/1) |> Enum.map_join("_", &String.upcase/1)}"

    default_value = Keyword.get(opts, :default_value)

    value =
      if System.get_env(env_variable) do
        System.get_env(env_variable)
      else
        get_in(secrets, keys)
      end

    if is_nil(value) do
      default_value
    else
      value
    end
  end

  def secrets do
    Application.get_env(:tuist, :secrets)[env()] || %{}
  end

  def put_application_secrets(secrets) do
    Application.put_env(:tuist, :secrets, secrets)
    :ok
  end

  @doc ~s"""
  It decrypts the secrets and returns them.
  """
  def decrypt_secrets() do
    if env() == :test do
      {:ok, secrets_map} =
        File.read!("priv/secrets/test_secrets.yml")
        |> YamlElixir.read_from_string()

      secrets_map
      |> to_atom_map()
    else
      master_key_path = Path.join("priv/secrets", "master.key")
      master_key_env_variable = "PHX_MASTER_KEY"
      secrets_path = System.get_env("SECRETS_PATH", "priv/secrets/secrets.yml.enc")

      if System.get_env(master_key_env_variable) || File.exists?(master_key_path) do
        key = System.get_env(master_key_env_variable) || File.read!(master_key_path)

        EncryptedSecrets.read!(key, secrets_path)
      else
        %{}
      end
    end
  end

  defp to_atom_map(map) when is_map(map) do
    Map.new(map, fn {k, v} -> {String.to_atom(k), to_atom_map(v)} end)
  end

  defp to_atom_map(value) do
    value
  end
end
